### ISSUE TITLE
Compilation warnings and errors when using Xcode 15 Beta

### ISSUE THREAD

**Issue by DenTelezhkin** [getsentry/sentry-cocoa#3122]
### Platform

iOS

### Installed

Swift Package Manager

### Version

8.8.0

### Steps to Reproduce

When compiling on device (compilation on simulator works fine), there are a bunch of new warnings, and with specific project configuration, compilation error when using Sentry.

**Update: Compilation error is fixed in Xcode 15 Beta 3.**

Warnings can be reproduced by doing the following:

1. Create new project
2. Add Sentry via SPM
3. Build for device

Observe following warnings:

```
Ignoring duplicate library '-lc++'

Duplicate -rpath '/Users/denystelezhkin/Library/Developer/Xcode/DerivedData/Sentry-Xcode15-acagjnxhvucyzkbtgzttahdkqjvl/Build/Products/Debug-iphoneos/PackageFrameworks' ignored

Could not associate debug note to atom l_OBJC_LABEL_PROTOCOL_$__TtP13SentryPrivate23SentryMXManagerDelegate_
Could not associate debug note to atom l_OBJC_LABEL_PROTOCOL_$_MXMetricManagerSubscriber
Could not associate debug note to atom l_OBJC_LABEL_PROTOCOL_$_NSObject
Could not associate debug note to atom l_OBJC_PROTOCOL_REFERENCE_$__TtP13SentryPrivate23SentryMXManagerDelegate_
Could not associate debug note to atom l_OBJC_PROTOCOL_REFERENCE_$_MXMetricManagerSubscriber
Could not associate debug note to atom l_OBJC_PROTOCOL_REFERENCE_$_NSObject
```

However, if project setup is as follows: App -> Framework -> Sentry, warnings are the same, but there is now a compile error:

```
Assertion failed: ((targetAddr % extras.instrArm64.scale) == 0), function applyFixup, file Fixup.cpp, line 850.
```

This is probably an Xcode beta issue, and may be mitigated in future betas of Xcode 15, however as this is reproducible in two beta releases already, I wanted to file this issue for visibility, as there might be something i'm missing here.

### Expected Result

Compilation succeeds

### Actual Result

Compliation failed.

### Are you willing to submit a PR?

Currently no, as I'm not certain how this issue can be fixed (and if this is a Sentry issue at all)

**Comment by brustolin** [comment]
Thanks @DenTelezhkin for pointing this out. We're going to investigate.

**Comment by hubertdeng123** [comment]
Removing the `Waiting for: Product Owner` label since this item looks like it is backlogged

**Comment by danmonaghan** [comment]
Interesting, I get a warning like this on another Swift package, [Pusher Web Socket issue](https://github.com/pusher/pusher-websocket-swift/issues/411), however mine doesn't fail compliation.

**Comment by DenTelezhkin** [comment]
Update: Compilation error is fixed in Xcode 15 Beta 3. There are a ton of new warnings though, a lot of them about Sentry, now present not only on iOS, but on macOS as well.

**Comment by philipphofmann** [comment]
I can reproduce the issue by compiling an empty sample project that includes Sentry via SPM and compiling it for a plugged-in device. As you stated @DenTelezhkin, they are now compiler warnings. We still have to investigate how we can fix these.

<details><summary>Build Warnings</summary>
<p>
```
ld: warning: ignoring duplicate library '-lc++'
ld: warning: duplicate -rpath '/Users/philipp.hofmann/Library/Developer/Xcode/DerivedData/SPM-gumhhehtvbtwtsapzzzrytpiduwa/Build/Products/Debug-iphoneos/PackageFrameworks' ignored
ld: warning: could not associate debug note to atom l_OBJC_LABEL_PROTOCOL_$__TtP13SentryPrivate23SentryMXManagerDelegate_
ld: warning: could not associate debug note to atom l_OBJC_LABEL_PROTOCOL_$_MXMetricManagerSubscriber
ld: warning: could not associate debug note to atom l_OBJC_LABEL_PROTOCOL_$_NSObject
ld: warning: could not associate debug note to atom l_OBJC_PROTOCOL_REFERENCE_$__TtP13SentryPrivate23SentryMXManagerDelegate_
ld: warning: could not associate debug note to atom l_OBJC_PROTOCOL_REFERENCE_$_MXMetricManagerSubscriber
ld: warning: could not associate debug note to atom l_OBJC_PROTOCOL_REFERENCE_$_NSObject
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_NSObject
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryIntegrationProtocol
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryANRTrackerDelegate
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentrySerializable
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentrySystemEventBreadcrumbsDelegate
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryCoreDataMiddleware
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryCrashBinaryImageProvider
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryCrashMachineContextWrapper
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryCrashReportFilter
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryScopeObserver
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryCurrentDateProvider
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryObjCRuntimeWrapper
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryRateLimits
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_NSCopying
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryTransport
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryEnvelopeRateLimitDelegate
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryFileManagerDelegate
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$__TtP13SentryPrivate23SentryMXManagerDelegate_
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_NSURLSessionDelegate
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryTracerDelegate
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryRequestManager
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryRandom
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryClientAttachmentProcessor
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentrySpan
ld: warning: could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryFramesTrackerListener

Ignoring duplicate library '-lc++'

Duplicate -rpath '/Users/philipp.hofmann/Library/Developer/Xcode/DerivedData/SPM-gumhhehtvbtwtsapzzzrytpiduwa/Build/Products/Debug-iphoneos/PackageFrameworks' ignored

Could not associate debug note to atom l_OBJC_LABEL_PROTOCOL_$__TtP13SentryPrivate23SentryMXManagerDelegate_
Could not associate debug note to atom l_OBJC_LABEL_PROTOCOL_$_MXMetricManagerSubscriber
Could not associate debug note to atom l_OBJC_LABEL_PROTOCOL_$_NSObject
Could not associate debug note to atom l_OBJC_PROTOCOL_REFERENCE_$__TtP13SentryPrivate23SentryMXManagerDelegate_
Could not associate debug note to atom l_OBJC_PROTOCOL_REFERENCE_$_MXMetricManagerSubscriber
Could not associate debug note to atom l_OBJC_PROTOCOL_REFERENCE_$_NSObject
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_NSObject
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryIntegrationProtocol
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryANRTrackerDelegate
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentrySerializable
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentrySystemEventBreadcrumbsDelegate
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryCoreDataMiddleware
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryCrashBinaryImageProvider
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryCrashMachineContextWrapper
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryCrashReportFilter
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryScopeObserver
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryCurrentDateProvider
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryObjCRuntimeWrapper
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryRateLimits
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_NSCopying
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryTransport
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryEnvelopeRateLimitDelegate
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryFileManagerDelegate
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$__TtP13SentryPrivate23SentryMXManagerDelegate_
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_NSURLSessionDelegate
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryTracerDelegate
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryRequestManager
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryRandom
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryClientAttachmentProcessor
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentrySpan
Could not associate debug note to atom __OBJC_LABEL_PROTOCOL_$_SentryFramesTrackerListener
```
</p>
</details>

**Comment by philipphofmann** [comment]
From my understanding, the warning message `"Ignoring duplicate library '-lc++'"` in Xcode indicates that the linker is encountering multiple occurrences of the same library during the build process, which doesn't make sense to me. This warning typically occurs when the same library is included or linked multiple times in your Xcode project. My best guess is that with Xcode 15, they changed something about how they link SPM packages. As the error changed to a warning, we will watch this over the subsequent Xcode releases. At the moment, I don't think there is much we can do. Do you have any ideas, @armcknight or @brustolin?

**Comment by brustolin** [comment]
I have a suspicion that I can't test right now.
Apple added C++ interoperability with Swift, which I think means they are adding C++ lib as default during compilation.
[We also add C++ lib in our `Package.swift`](https://github.com/getsentry/sentry-cocoa/blob/main/Package.swift#L37)

You may try to remove this from the package to see if it works. I would call this an Apple bug, they definitely should check for duplicity.

**Comment by DenTelezhkin** [comment]
@brustolin I tried removing linker settings in my fork (https://github.com/DenTelezhkin/sentry-cocoa/tree/feature/xcode-15-spm-support), unfortunately, it did not help, warnings remain. I had a thought, that maybe because C++ language standard is included in Package.swift, Sentry and Xcode compiler use different C++ language versions, which causes a conflict. However, i cannot test this theory, as removing `cxxLanguageStandard` from Package.swift breaks Sentry compilation.

And I'm with you, definitely seems like an Apple bug at this moment.

**Comment by mezhevikin** [comment]
He guys. I dont use sentry-cocoa but I have the same waning in my project in Xcode 15 beta 3. I thing it is beta problem

**Comment by waydabber** [comment]
The issue persists in XCode 15 Beta 3.

**Comment by philipphofmann** [comment]
@mezhevikin and @waydabber, do you get warnings or build errors?

We are aware that they are warnings, but as pointed out in https://github.com/getsentry/sentry-cocoa/issues/3122#issuecomment-1623703806 and https://github.com/getsentry/sentry-cocoa/issues/3122#issuecomment-1623533547, we assume the warnings could be bugs from Apple, and they might be resolved in future Xcode releases.

**Comment by philipphofmann** [comment]
Warnings are still there for Xcode 15 Beta 4.

**Comment by AlexHadfield** [comment]
<img width="366" alt="Xcode Beta Build Errors" src="https://github.com/getsentry/sentry-cocoa/assets/25065732/bb98a8a0-1b22-48bf-9952-a04d800878ab">

These are the Sentry build errors I am seeing when trying to build a project in Xcode 15 Beta 4.  It builds fine to the simulator, then fails when trying to build and install on a physical device.  It might get resolved in a later beta.  Just wanted to note it here.

**Comment by jmanalang3** [comment]
This warning showing on Xcode beta 4.

**Comment by AlexHadfield** [comment]
This is still happening in Xcode Version 15.0 beta 5 (15A5209g)

**Comment by DenTelezhkin** [comment]
Well, actually on Xcode 15 beta 5 much of the warnings have disappeared. There are only two warnings left:

```
Ignoring duplicate library '-lc++'
Ignoring duplicate library '-lz'
```
They appear in every binary Sentry is imported into.

**Comment by brustolin** [comment]
I could not reproduce `Ignoring duplicate library '-lz'` but I did get rid of `Ignoring duplicate library '-lc++' ` by removing this lines from `Package.swift`:

```
            linkerSettings: [
                .linkedLibrary("z"),
                .linkedLibrary("c++")
            ]
```

And I believe this will also solve the `-lz` duplication

**Comment by AlexHadfield** [comment]
You guys aren't seeing this anymore? It's still appearing when I try and run on a physical device: 
<img width="343" alt="Screenshot 2023-07-27 at 10 27 48 AM" src="https://github.com/getsentry/sentry-cocoa/assets/25065732/236637d5-dce8-4093-92e6-dfbf0ec6130f">
<img width="642" alt="Screenshot 2023-07-27 at 10 27 34 AM" src="https://github.com/getsentry/sentry-cocoa/assets/25065732/5ffa6df4-2bc5-4af6-aacf-b7dbff752e9a">

**Comment by brustolin** [comment]
Hello @AlexHadfield, I just noticed this problem is for WatchOS. 
I haven't try it yet. I will take a look next.

**Comment by soucolline** [comment]
@brustolin is this fix only for SPM ? Using cocoapods i'm still seeing the warning

**Comment by brustolin** [comment]
@soucolline You're right, this was not changed for cocoapod yet. I will see if it is possible.

**Comment by brustolin** [comment]
@soucolline this was fixed in #3312 and will be available in our next release.

**Comment by soucolline** [comment]
Thanks a lot for keeping me updated @brustolin
